File: Solutions/Python/minimum_operations_to_make_a_uni_value_grid.py
"""
    [MEDIUM]
    2033. Minimum Operations to Make a Uni-Value Grid

    Concepts:
    - array
    - math

    Stats:
        Runtime | 192 ms    [Beats 38.16%]
        Memory  | 39.28 MB  [Beats 36.40%]
"""

class Solution:
    def minOperations(self, grid: List[List[int]], x: int) -> int:
        values = sorted([val for row in grid for val in row])
        n = len(values)

        if n == 1:
            return 0

        if not all([val % x == values[0] % x for val in values]):
            return -1

        return sum([abs(values[n // 2] - val) // x for val in values])

File: Solutions/Rust/minimum_operations_to_make_a_uni_value_grid.rs
/*
    [MEDIUM]
    2033. Minimum Operations to Make a Uni-Value Grid

    Concepts:
    - array
    - math

    Stats:
        Runtime | 14 ms     [Beats 50.00%]
        Memory  | 8.55 MB   [Beats 25.00%]
*/
impl Solution {
    pub fn min_operations(grid: Vec<Vec<i32>>, x: i32) -> i32 {
        let mut values = grid
            .into_iter()
            .flatten()
            .collect::<Vec<i32>>();
        values.sort();
        let n = values.len();

        if n == 1 {
            return 0;
        }

        if !values.iter().all(|&val| (val % x) == (values[0] % x)) {
            return -1;
        }

        return values
            .iter()
            .map(|&val| (values[n / 2] - val).abs() / x)
            .fold(0, |a, b| a + b);
    }
}